examples/resource_partitioner/async_customization.cpp

```diff
@@ -0,0 +1,534 @@
+    // --------------------------------------------------------------------
+    template <typename F, typename ... Ts>
+    future<typename util::invoke_result<F, Ts...>::type>
+    async_execute(F && f, Ts &&... ts)
+    {
+        typedef typename util::detail::invoke_deferred_result<F, Ts...>::type
+            result_type;
+
+        using namespace hpx::util::debug;
+        std::cout << "async_execute : Function    : "
+                  << print_type<F>() << "\n";
+        std::cout << "async_execute : Arguments   : "
+                  << print_type<Ts...>(" | ") << "\n";
+        std::cout << "async_execute : Result      : "
+                  << print_type<result_type>() << "\n";
+
+        // forward the task execution on to the real internal executor
+        lcos::local::futures_factory<result_type()> p(
+            executor_,
+            util::deferred_call(std::forward<F>(f),
+                                     std::forward<Ts>(ts)...)
+        );
+
+        p.apply(
+            launch::async,
+            threads::thread_priority_default,
+            threads::thread_stacksize_default);
+
+        return p.get_future();
+    }
+
+    // --------------------------------------------------------------------
+    // .then() execute specialized for a future<P> predecessor argument
+    // note that future<> and shared_future<> are both supported
+    // --------------------------------------------------------------------
+    template <typename F,
+              typename Future,
+              typename ... Ts,
+              typename = typename std::enable_if_t<traits::is_future<
+              typename std::remove_reference<Future>::type>::value>>
+    auto
+    then_execute(F && f, Future&& predecessor, Ts &&... ts)
+    ->  future<typename util::detail::invoke_deferred_result<
+        F, Future, Ts...>::type>
+    {
+        typedef typename util::detail::invoke_deferred_result<
+                F, Future, Ts...>::type result_type;
+
+        using namespace hpx::util::debug;
+        std::cout << "then_execute : Function     : "
+                  << print_type<F>() << "\n";
+        std::cout << "then_execute : Predecessor  : "
+                  << print_type<Future>() << "\n";
+        std::cout << "then_execute : Future       : "
+                  << print_type<typename traits::future_traits<Future>::result_type>()
+                  << "\n";
+        std::cout << "then_execute : Arguments    : "
+                  << print_type<Ts...>(" | ") << "\n";
+        std::cout << "then_execute : Result       : "
+                  << print_type<result_type>() << "\n";
+
+        // forward the task on to the 'real' underlying executor
+        lcos::local::futures_factory<result_type()> p(
+            executor_,
+            util::deferred_call(std::forward<F>(f),
+                                std::forward<Future>(predecessor),
+                                std::forward<Ts>(ts)...)
+        );
```

Comment:
This will execute the passed function right away, right? There is no ".then" semantic here, I think.